WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/ranges/data/AnalyticsHubMonthToDateRangeData.kt

```diff
@@ -0,0 +1,40 @@
+package com.woocommerce.android.ui.analytics.ranges.data
+
+import com.woocommerce.android.extensions.oneMonthAgo
+import com.woocommerce.android.extensions.startOfCurrentMonth
+import com.woocommerce.android.ui.analytics.ranges.AnalyticsHubTimeRange
+import com.woocommerce.android.ui.analytics.ranges.AnalyticsHubTimeRangeData
+import java.util.Calendar
+import java.util.Date
+
+// Responsible for defining two ranges of data, the current one, starting from the first day of the current month
+// until the current date, and the previous one, starting from the first day of the previous month
+// until the same day of the previous month. E. g.
+//
+// Today: 31 Jul 2022
+// When user requests report at 05:49 PM
+// Current range: Jul 1, 00:00 until Jul 31, 05:49 PM, 2022
+// Previous range: Jun 1, 00:00 until Jun 30, 05:49 PM, 2022
+//
+class AnalyticsHubMonthToDateRangeData(
+    referenceDate: Date,
+    calendar: Calendar
+) : AnalyticsHubTimeRangeData {
+    override val currentRange: AnalyticsHubTimeRange
+    override val previousRange: AnalyticsHubTimeRange
+
+    init {
+        calendar.time = referenceDate
```

Comment:
Similar concern as with `CalendarExt` - before starting modifications on `calendar`, I'd clone it to a local object to lower the chances of unwanted state mutation.

In other words: when I pass `Calendar` to `AnalyticsHubMonthToDateRangeData` I don't expect that my `Calendar` will change.

```suggestion
    init {
        val calendarClone = calendar.clone as Calendar // or any other name, just a sample 👍 
        calendarClone.time = referenceDate
```

Reply:
Again, thanks for catching this calendar issue. It could lead to dangerous problems! I've adjusted the base class to enforce this copy to all date range classes in this commit: https://github.com/woocommerce/woocommerce-android/pull/8063/commits/04294bcfb49d5f53d25834ed9318418b152ce933

Let me know what you think of this approach!

Reply:
Nice idea with the base class! 👏